### PR TITLE
Вершинина Александра. Задача 1. Вариант 21. Интегрирование – метод Монте-Карло

### PR DESCRIPTION
**Pull request by sanya2905**
Последовательное решение:
Определяем сколько случайных точек попало в нашу область, вычисляем плотность и с помощью нее получаем значение нужного нам интеграла.
Параллельное решение:
Тоже самое, только используется фиксированное число итераций, входные данные рассылаются всем процессам через boost::mpi::broadcast и распараллеливается финальная часть алгоритма.

### CODE REVIEW

**Review by ermilovad — CHANGES_REQUESTED** [review]
Use std::copy

**Comment by sanya2905** [comment]
> Use std::copy

fixed :)

**Review by ermilovad — APPROVED** [review]

**Review by belovtut — CHANGES_REQUESTED** [review]
You specify the number of iterations manually twice, so when you want to change the number, you have to change both lines. You may use constant

**Comment by sanya2905** [comment]
> You specify the number of iterations manually twice, so when you want to change the number, you have to change both lines. You may use constant

fixed :)

**Review by belovtut — APPROVED** [review]

File: tasks/mpi/vershinina_a_integration_the_monte_carlo_method/include/ops_mpi.hpp [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+#include <numeric>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "core/task/include/task.hpp"
+
+namespace vershinina_a_integration_the_monte_carlo_method {
+
+std::vector<double> getRandomVector(double sz);
```
Comment (aobolensk) [inline tasks/mpi/vershinina_a_integration_the_monte_carlo_method/include/ops_mpi.hpp:17]: Please, move it to test module

File: tasks/seq/vershinina_a_integration_the_monte_carlo_method/func_tests/main.cpp [hunk 2]
```diff
@@ -0,0 +1,65 @@
+#include <gtest/gtest.h>
+
+#include <cmath>
+#include <vector>
+
+#include "seq/vershinina_a_integration_the_monte_carlo_method/include/ops_seq.hpp"
+
+TEST(vershinina_a_integration_the_monte_carlo_method, test1) {
```
Comment (aobolensk) [inline tasks/seq/vershinina_a_integration_the_monte_carlo_method/func_tests/main.cpp:8]: Please, add random tests for seq as well

File: tasks/mpi/vershinina_a_integration_the_monte_carlo_method/perf_tests/main.cpp [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+TEST(vershinina_a_integration_the_monte_carlo_method, test_pipeline_run) {
+  boost::mpi::communicator world;
+  std::vector<double> in;
+  std::vector<double> global_res(1, 0);
+
+  std::shared_ptr<ppc::core::TaskData> taskDataPar = std::make_shared<ppc::core::TaskData>();
+  double count_size_vector;
+
+  if (world.rank() == 0) {
+    count_size_vector = 4;
```
Comment (aobolensk) [inline tasks/mpi/vershinina_a_integration_the_monte_carlo_method/perf_tests/main.cpp:19]: Perf test should be run on larger sizes

**Review by aobolensk — CHANGES_REQUESTED** [review]

File: tasks/mpi/vershinina_a_integration_the_monte_carlo_method/perf_tests/main.cpp (hunk 3 above)
Comment (sanya2905) [reply tasks/mpi/vershinina_a_integration_the_monte_carlo_method/perf_tests/main.cpp:19]: > Perf test should be run on larger sizes

In my solution the vector can consist of only 4 values: xmin xmax (integration limits), ymin ymax (function limits)

File: tasks/seq/vershinina_a_integration_the_monte_carlo_method/func_tests/main.cpp (hunk 2 above)
Comment (sanya2905) [reply tasks/seq/vershinina_a_integration_the_monte_carlo_method/func_tests/main.cpp:8]: > Please, add random tests for seq as well

What should be used as a reference solution?

File: tasks/mpi/vershinina_a_integration_the_monte_carlo_method/include/ops_mpi.hpp (hunk 1 above)
Comment (sanya2905) [reply tasks/mpi/vershinina_a_integration_the_monte_carlo_method/include/ops_mpi.hpp:17]: > Please, move it to test module

fixed :)

File: tasks/seq/vershinina_a_integration_the_monte_carlo_method/src/ops_seq.cpp [hunk 4]
```diff
@@ -0,0 +1,56 @@
+#include "seq/vershinina_a_integration_the_monte_carlo_method/include/ops_seq.hpp"
+
+#include <algorithm>
+#include <functional>
+#include <random>
+#include <string>
+#include <thread>
+#include <vector>
+
+using namespace std::chrono_literals;
```
Comment (allnes) [inline tasks/seq/vershinina_a_integration_the_monte_carlo_method/src/ops_seq.cpp:10]: please remove it

File: tasks/mpi/vershinina_a_integration_the_monte_carlo_method/src/ops_mpi.cpp [hunk 5]
```diff
@@ -0,0 +1,125 @@
+#include "mpi/vershinina_a_integration_the_monte_carlo_method/include/ops_mpi.hpp"
+
+#include <algorithm>
+#include <functional>
+#include <random>
+#include <string>
+#include <thread>
+#include <vector>
+
+using namespace std::chrono_literals;
```
Comment (allnes) [inline tasks/mpi/vershinina_a_integration_the_monte_carlo_method/src/ops_mpi.cpp:10]: please remove it

File: tasks/seq/vershinina_a_integration_the_monte_carlo_method/src/ops_seq.cpp [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  return taskData->inputs_count[0] == 4 && taskData->outputs_count[0] == 1;
+}
+
+bool vershinina_a_integration_the_monte_carlo_method::TestTaskSequential::run() {
+  internal_order_test();
+  int count;
+  double total = 0;
+  double inBox = 0;
+  reference_res = 0;
+  for (count = 0; count < 100000; count++) {
```
Comment (allnes) [inline tasks/seq/vershinina_a_integration_the_monte_carlo_method/src/ops_seq.cpp:33]: please parametrize iterations count

**Review by allnes — CHANGES_REQUESTED** [review]

File: tasks/seq/vershinina_a_integration_the_monte_carlo_method/func_tests/main.cpp (hunk 2 above)
Comment (aobolensk) [reply tasks/seq/vershinina_a_integration_the_monte_carlo_method/func_tests/main.cpp:8]: Integral value that is can be calculated using math formula

File: tasks/mpi/vershinina_a_integration_the_monte_carlo_method/perf_tests/main.cpp (hunk 3 above)
Comment (aobolensk) [reply tasks/mpi/vershinina_a_integration_the_monte_carlo_method/perf_tests/main.cpp:19]: Yes, but you need to be able to customize iterations count.

See https://github.com/learning-process/ppc-2024-autumn/pull/258#discussion_r1835557138 as well

File: tasks/seq/vershinina_a_integration_the_monte_carlo_method/func_tests/main.cpp (hunk 2 above)
Comment (sanya2905) [reply tasks/seq/vershinina_a_integration_the_monte_carlo_method/func_tests/main.cpp:24]: > Integral value that is can be calculated using math formula

fixed :)

File: tasks/seq/vershinina_a_integration_the_monte_carlo_method/src/ops_seq.cpp (hunk 4 above)
Comment (sanya2905) [reply tasks/seq/vershinina_a_integration_the_monte_carlo_method/src/ops_seq.cpp]: > please remove it

fixed :)

File: tasks/mpi/vershinina_a_integration_the_monte_carlo_method/src/ops_mpi.cpp (hunk 5 above)
Comment (sanya2905) [reply tasks/mpi/vershinina_a_integration_the_monte_carlo_method/src/ops_mpi.cpp]: > please remove it

fixed :)

File: tasks/seq/vershinina_a_integration_the_monte_carlo_method/src/ops_seq.cpp (hunk 6 above)
Comment (sanya2905) [reply tasks/seq/vershinina_a_integration_the_monte_carlo_method/src/ops_seq.cpp]: > please parametrize iterations count

fixed :)

File: tasks/mpi/vershinina_a_integration_the_monte_carlo_method/perf_tests/main.cpp (hunk 3 above)
Comment (sanya2905) [reply tasks/mpi/vershinina_a_integration_the_monte_carlo_method/perf_tests/main.cpp]: > Yes, but you need to be able to customize iterations count.
> 
> See [#258 (comment)](https://github.com/learning-process/ppc-2024-autumn/pull/258#discussion_r1835557138) as well

fixed :)

**Review by allnes — APPROVED** [review]

**Pull request merged**